From timur.kristof@gmail.com Fri Jun  1 03:33:07 2018
Return-Path: <timur.kristof@gmail.com>
Received: from localhost.localdomain
 ([2a02:ab88:63c0:1a00:9eb6:d0ff:fe89:a491]) by smtp.gmail.com with ESMTPSA
 id 67-v6sm1715790wmw.32.2018.06.01.03.33.06 (version=TLS1_2
 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256); Fri, 01 Jun 2018 03:33:07
 -0700 (PDT)
From: "=?UTF-8?q?Timur=20Krist=C3=B3f?=" <timur.kristof@gmail.com>
X-Google-Original-From: =?UTF-8?q?Timur=20Krist=C3=B3f?=
 <venemo@fedoraproject.org>
To: timur.kristof@gmail.com,
	platform-driver-x86@vger.kernel.org
Cc: =?UTF-8?q?Timur=20Krist=C3=B3f?= <venemo@fedoraproject.org>,
	mjg59@srcf.ucam.org,
	pali.rohar@gmail.com,
	dvhart@infradead.org,
	andy@infradead.org
Subject: [PATCH v2] dell-laptop: fix AC keyboard backlight timeout on Dell
 XPS 13 9370
Date: Fri,  1 Jun 2018 12:32:56 +0200
Message-Id: <20180601103256.4571-1-venemo@fedoraproject.org>
X-Mailer: git-send-email 2.17.0
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
X-Evolution-Source: 572fc47725462d0a763b181e96ce3f766af355fa
Content-Transfer-Encoding: 8bit

The 9370 doesn't expose the necessary KBD_LED_AC_TOKEN in the
BIOS, so the driver thinks it cannot adjust the AC keyboard
backlight timeout. This patch adds a quirk to fix this until
Dell adds the missing token to the BIOS.

For further discussion, see:
https://github.com/dell/libsmbios/issues/48

Signed-off-by: Timur Kristóf <venemo@fedoraproject.org>
---
 drivers/platform/x86/dell-laptop.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/drivers/platform/x86/dell-laptop.c b/drivers/platform/x86/dell-laptop.c
index c52c6723374b..058944258161 100644
--- a/drivers/platform/x86/dell-laptop.c
+++ b/drivers/platform/x86/dell-laptop.c
@@ -38,6 +38,7 @@
 struct quirk_entry {
 	bool touchpad_led;
 	bool kbd_led_levels_off_1;
+	bool kbd_missing_ac_tag;
 
 	bool needs_kbd_timeouts;
 	/*
@@ -68,6 +69,10 @@ static struct quirk_entry quirk_dell_xps13_9333 = {
 	.kbd_timeouts = { 0, 5, 15, 60, 5 * 60, 15 * 60, -1 },
 };
 
+static struct quirk_entry quirk_dell_xps13_9370 = {
+	.kbd_missing_ac_tag = true,
+};
+
 static struct quirk_entry quirk_dell_latitude_e6410 = {
 	.kbd_led_levels_off_1 = true,
 };
@@ -291,6 +296,15 @@ static const struct dmi_system_id dell_quirks[] __initconst = {
 		},
 		.driver_data = &quirk_dell_xps13_9333,
 	},
+	{
+		.callback = dmi_matched,
+		.ident = "Dell XPS 13 9370",
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "Dell Inc."),
+			DMI_MATCH(DMI_PRODUCT_NAME, "XPS 13 9370"),
+		},
+		.driver_data = &quirk_dell_xps13_9370,
+	},
 	{
 		.callback = dmi_matched,
 		.ident = "Dell Latitude E6410",
@@ -1401,7 +1416,8 @@ static inline int kbd_init_info(void)
 	 *       timeout value which is shared for both battery and AC power
 	 *       settings. So do not try to set AC values on old models.
 	 */
-	if (dell_smbios_find_token(KBD_LED_AC_TOKEN))
+	if ((quirks && quirks->kbd_missing_ac_tag) ||
+		 dell_smbios_find_token(KBD_LED_AC_TOKEN))
 		kbd_timeout_ac_supported = true;
 
 	kbd_get_state(&state);
-- 
2.17.0
(quirks && quirks->kbd_missing_ac_tag)

